scipy/optimize/_optimize.py

```diff
@@ -3477,12 +3489,23 @@ def brute(func, ranges, args=(), Ns=20, full_output=0, finish=fmin,
     for k in range(N):
         if type(lrange[k]) is not type(slice(None)):
             if len(lrange[k]) < 3:
-                lrange[k] = tuple(lrange[k]) + (complex(Ns),)
+                if integrality[k]:
+                    ub = np.floor(lrange[k][1])
+                    lb = min(np.ceil(lrange[k][0]), ub)
+                    lrange[k] = lb, ub
+                    n_ints_between_bounds = lrange[k][1] - lrange[k][0] + 1
+                    n_points = min(Ns, n_ints_between_bounds)
+                else:
+                    n_points = Ns
+                lrange[k] = tuple(lrange[k]) + (complex(n_points),)
             lrange[k] = slice(*lrange[k])
     if (N == 1):
         lrange = lrange[0]
 
     grid = np.mgrid[lrange]
+    for k in range(N):
+        if integrality[k]:
+            grid[k, ...] = np.round(grid[k, ...])
```

Comment:
are there any circumstances where this line would result in an integer being outside the `(low, high)` values originally supplied?

Reply:
Yes, I think there are. If the user supplies a `slice` to start with, then this is possible. For example, if a user supplies `ranges=[slice(0.4, 0.9, 21j)], integrality=True`

Then:

1. There will be integer values of 0 and 1 being produced in the grid.
2. The rounded values of 0 and 1 are outside the original slice.
3. There should be no integers in this slice (and an Exception should be raised).

Reply:
Slices give the user the opportunity to select the exact points at which the function is to be evaluated. For example, it allows the user to choose their own integers. 
How about just stating that if the user supplies a slice, then `integrality` will be ignored for that input?

Reply:
Having now realised fully that slices allow the user to specify integer grids I wonder if it's worth adding integrality at all?
That being said, it probably is harder to think about creating the correct slice, and having `integrality` only enforce it over the `Ns` generated points, as per your suggestion, is probably the best way of doing it.

As an aside, thinking about fitting discrete distributions, or `brute` optimisation in general - there's a reasonable probability that the integers corresponding to the best solution won't get explored on a single pass.

Reply:
I would still check for the presence of at least one integer though

Reply:
One problem is that `brute` (in `main`) errors out if there is only one value in the grid. It also "works" - giving an out-of-bounds answer - when the original `range` shouldn't contain any points. It's tough to tiptoe around these bugs when `brute` would be so simple to write from scratch. I'd suggest that it is enough for `integrality` to just not introduce new bugs. Otherwise, I think it would be better to just write `brute` from scratch in a more modern style.